Update README.md (#10)
RoboTeamTwente/roboteam_suite: README.md

# roboteam_suite
This repo contains:

> - Roboteam AI, which listens to Roboteam World and makes decisions on what the robots should do based on the vision data.

Reviewer: And robothub

Reviewer 2: Hub receives feedback but I'm not sure if that's redirected to AI directly or goes through world first

Reviewer: directly to Ai, we're planning to change this.

- Roboteam World, which interprets vision data (received from GRSim, or from a physical camera)
- Roboteam Robothub, used to send commands to the robots (either through GRSim or the base station)
- Roboteam Utils, which contains a lot of helper functions, primarily geometrical in nature, such as Lines, Triangles, etc. These are used in other repo's, such as Roboteam AI and Roboteam World
- Roboteam Proto, which contains the protocol used to send and receive messages over TCP/UDP 

In order to test this code, we use the GRSim simulator developed by the SSL Team Parsian. However, due to differences in the control mechanism of the robot (specifically, theta vs omega control), this software requires the [RoboTeam fork of GRSim](https://github.com/RoboTeamTwente/grSim). Installation instructions can be found in the readme of that repo.

Please note:

> This software is tested on Ubuntu 18.04 and 20.04 LTS, and the installation instructions are written with that distro/OS in mind. However, other distro's of Linux are also likely to be able to run this software, albeit with some modifications.

Reviewer 3: Maybe just mention ubuntu 20 since Travis now only runs Ubuntu 20?

> This software is tested on Ubuntu 18.04 and 20.04 LTS, and the installation instructions are written with that distro/OS in mind. However, other distro's of Linux are also likely to be able to run this software, albeit with some modifications.

Reviewer 2: ```suggestion
This software is tested on Ubuntu 20.04 LTS, and the installation instructions are written with that distro/OS in mind. However, other distro's of Linux are also likely to be able to run this software, albeit with some modifications.
```

# Usage

> First, clone this repository. We recommed using an SSH key if you are able to do so, as this immediately authorizes your future pushes to the repositories. 

Reviewer 3: ```suggestion
First, clone this repository. We recommend using an SSH key if you are able to do so, as this immediately authorizes your future pushes to the repositories. 
```

```
git clone --recurse-submodule git@github.com:RoboTeamTwente/roboteam_suite.git
```
Now, you can open the roboteam_suite project with an IDE of your choice. We recommed CLion, as it has a lot of very helpful features and you can get a student license for it. 

> Make sure you have CMake installed. Go the CMakeLists.txt and click on it. CLion will generate a pop-up saying "project files not loaded" in blue in the top right corner. Clicking it will have CMake load the files. This means it generates instructions on how the files should be compiled in order to produce the executables specified in the CMakeLists.txt file. You can now compile the targets!

Reviewer 3: ```suggestion
Make sure you have CMake installed. Go to the CMakeLists.txt and click on it. CLion will generate a pop-up saying "project files not loaded" in blue in the top right corner. Clicking it will have CMake load the project. This means it generates instructions on how the files should be compiled in order to produce the executables specified in the CMakeLists.txt file. You can now compile the targets!
```

The easiest way to run the application is to make some [compound executables](https://www.jetbrains.com/help/clion/run-debug-configuration.html#config-folders) in CLion that boot everything you need. To run 1 team in GRSim, the simulator, you will need to add roboteam_ai, roboteam_world, and roboteam_robothub. 
Make sure you have GRSim or ssl-vision running and are listening to the right ports. In general, the GRSim vision multicast port should be set to 10006.

## Auxiliary Software
It is recommended to use the game controller to control the gamestates. You can do so by downloading the latest binary release of the [SSL-Game-Controller](https://github.com/RoboCup-SSL/ssl-game-controller) repo, and executing it. For example, when downloaded to the Downloads folder:
```
~/Downloads/ssl-game-controller_v1.2.0_linux_amd64
```

> You might need to chmod the file to make it executable. The AI client should now listen to the commands sent by the GameController. 

Reviewer 3: Maybe also mention the chmod command

> You might need to chmod the file to make it executable. The AI client should now listen to the commands sent by the GameController. 

Reviewer 2: ```suggestion
You might need to [chmod](https://www.howtoforge.com/tutorial/linux-chmod-command/) the file to make it executable. The AI client should now listen to the commands sent by the GameController. 
```

Also, make sure you have installed GRSim using the instructions found on the [RoboTeam fork of GRSim](https://github.com/RoboTeamTwente/grSim). Now execute the grsim binary from the command line. This usually looks like this:
```
/directory_you_cloned_grsim/grSim/bin/grsim
```
## Making the executables: 

Make sure the vision multicast port is set to 10006

### One Team: 
Make the compound executable shown below:
<img src="https://github.com/RoboTeamTwente/roboteam_suite/blob/RobotJesse-patch-1/readme_images/run_two_teams.png?raw=true" width="800" height="500"><br>

> Then run this with grsim on. 

Reviewer 3: ```suggestion
Then run this compound executable and run GRSim. 
```

> <img src="https://github.com/RoboTeamTwente/roboteam_suite/blob/RobotJesse-patch-1/readme_images/run_two_teams.png?raw=true" width="800" height="500"><br>

Reviewer: Uhhm, there's a markdown way to do this.
```suggestion
![](https://github.com/RoboTeamTwente/roboteam_suite/blob/RobotJesse-patch-1/readme_images/run_two_teams.png?raw=true)
```

Reviewer 2: That doesn't let you specify size and it makes the image humongous

Reviewer: Fair enough, mb

### Two Teams:
Follow the steps used for the other compound executable, only now add 2 extra targets: roboteam_ai_1 and roboteam_robothub_1, as seen in the screenshot, to the compound. While creating the new CMake Targets, be sure to include a "1" in the program arguments, as seen in the image.
<img src="https://github.com/RoboTeamTwente/roboteam_suite/blob/RobotJesse-patch-1/readme_images/run_two_teams.png?raw=true" width="800" height="500"><br>

> # Installation

Reviewer 3: Looks like you are missing googleTest/googleMock as a dependency

> <img src="https://github.com/RoboTeamTwente/roboteam_suite/blob/RobotJesse-patch-1/readme_images/run_two_teams.png?raw=true" width="800" height="500"><br>

Reviewer: ```suggestion
![](https://github.com/RoboTeamTwente/roboteam_suite/blob/RobotJesse-patch-1/readme_images/run_two_teams.png?raw=true)
```

# roboteam_suite
This repo contains:

> - Roboteam AI, which listens to Roboteam World and makes decisions on what the robots should do based on the vision data.

Reviewer 2: ```suggestion
- Roboteam AI, which makes decisions on what the robots should do based on the vision data, and feedback data (received from robots) which it receives from Roboteam World and Roboteam Robothub respectively.
```

SSL-Vision (https://github.com/RoboCup-SSL/ssl-vision)

Working with a referee
[SSL-Game-controller](https://github.com/RoboCup-SSL/ssl-game-controller)
[SSL-vision-client](https://github.com/RoboCup-SSL/ssl-vision-client)

Working with the grSim simulator
[grSim](https://github.com/RoboTeamTwente/grSim) 

# Usage
Easiest way to run the application is to make some [compound executables](https://www.jetbrains.com/help/clion/run-debug-configuration.html#config-folders) in CLion that boot everything you need. To run 1 team in GRSim, the simulator, you will need to add roboteam_ai, roboteam_world, and roboteam_robothub. 

> Make sure you have grsim or ssl-vision running and are listening to the right ports. In general, the GRSim vision multicast port should be set to 10006.

Reviewer 3: Maybe also add the compound setup for 2 teams

# Installation
## List of dependencies

- [CMake 3.10](https://cmake.org/)
- [Protobuf 3.9.1](https://developers.google.com/protocol-buffers/)
- [ZeroMQ 4.2.5](https://zeromq.org/)
- [zmqpp 4.1.2](https://github.com/zeromq/zmqpp) 

> - [Armadillo 9.6](http://arma.sourceforge.net/)

Reviewer 3: I don't think I had to install armadillo
